### PR TITLE
Reworked traceback

### PR DESCRIPTION
**Pull request by adamnovak**
This PR brings in my new traceback code, which saves the E and F matrices and uses them during the traceback.

It also includes @AlexKuhnle's changes introducing 'X' and 'N' CIGAR operations, and adding file arguments to some of the print methods.

Closes #1.

### CODE REVIEW

File: src/gssw.c [hunk 1]
```diff
@@ -945,58 +1230,193 @@ gssw_cigar* gssw_alignment_trace_back_word (gssw_align* alignment,
...
-            h = u;
-            --j;
+    while (LIKELY(scoreHere > 0 && i >= 0 && j >= 0)) {
+        // We're not out of score, and we're not off the left or top of the matrix
+    
+        // We think differently than we did creating the matrix. When creating
+        // the matrix, we care about the relative scores, because we want to
+        // come from the places giving the best score.
+        
+        // When computing the traceback, the scores are already determined, and
```
Comment (ekg) [inline src/gssw.c:1277]: This is spot on.

File: src/gssw.h [hunk 2]
```diff
@@ -22,6 +22,47 @@
...
+// XMM7, as assigned by the compiler. In C, these values have type "__m128i",
+// with the "i" meaning integer (because we do integer math on them). We have
+// code to use these in alignments, either to store 16 single-byte values or 8
+// double-byte values, depending on how large we're allowing the scores to get.
+
+// We use j for indexes in the read (or "query"), and i for indexes in the
+// reference (or "database"). Conceptually, the database runs rightwards along
+// the top of the DP matrix, and the read runs downwards along the left.
+
+// Note that this usage of i and j is BACKWARDS from the paper!
```
Comment (ekg) [inline src/gssw.h:36]: I assume you mean the SSW paper?

Also, we should write this up. @adamnovak I think this work rationalizes gssw and completes the project; pending some testing via vg.

**Comment by ekg** [comment]
@adamnovak this is awesome work. You have rescued GSSW from a confused place and brought the project very near to completion. We should think about writing this up as an application note or at least a preprint. It would only need to be a few pages as we could limit ourselves to a description of the generalization of SSW to POA.

**Comment by adamnovak** [comment]
More papers are always better.

For a paper, we should probably collect some speed comparisons with the
original, to see what sort of impact the dumping of the matrices actually
has. If we really are concerned about speed, we might consider dumping the
striped matrices, and unscrambling later (or even never if we write code to
find the correct elements to read).

**Pull request merged**